src/examples/libpmemobj/hashmap/hashmap_rp.c

```diff
@@ -0,0 +1,700 @@
+	PMEMoid value)
+{
+	assert(hashmap->count + 1 < hashmap->resize_threshold);
+
+	struct pobj_action actv[HASHMAP_RP_MAX_ACTIONS];
+
+	struct add_entry args;
+	args.data.key = key;
+	args.data.value = value;
+	args.data.hash = hash(hashmap, key);
+	args.pos = args.data.hash;
+	args.actv = actv;
+	args.actv_cnt = 0;
+
+	int dist = 0;
+	struct entry *entry_p = NULL;
+
+	int swaps = 0;
+
+
+	for (int n = 0; n < HASHMAP_RP_MAX_SWAPS; ++n) {
+		entry_p = (struct entry *)pmemobj_direct(hashmap->entries.oid);
+		entry_p += args.pos;
+
+		if (hashmap_rp_debug)
+			args.swaps = swaps;
+
+		/* Case 1: key already exists, override value */
+		if (entry_p->key == args.data.key) {
+			entry_update(pop, hashmap, &args);
+			pmemobj_publish(pop, args.actv, args.actv_cnt);
+
+			return 1;
+		}
+
+		/* Case 2: slot is empty from the beginning */
+		if (entry_p->hash == 0) {
+			entry_add(pop, hashmap, &args);
+			pmemobj_publish(pop, args.actv, args.actv_cnt);
+
+			return 0;
+		}
+
+		/*
+		 * Case 3: existing element (or tombstone) has probed less than
+		 * current element. Swap them (or put into tombstone slot) and
+		 * keep going to find another slot for that element.
+		 */
+		int existing_dist = probe_distance(hashmap, entry_p->hash,
+						args.pos);
+		if (existing_dist < dist) {
+			if (entry_is_deleted(entry_p->hash)) {
+				entry_add(pop, hashmap, &args);
+				pmemobj_publish(pop, args.actv, args.actv_cnt);
+
+				return 0;
+			}
+
+			struct entry temp;
+			temp.key = entry_p->key;
+			temp.value = entry_p->value;
+			temp.hash = entry_p->hash;
+
+			entry_update(pop, hashmap, &args);
+
+			args.data.key = temp.key;
+			args.data.value = temp.value;
+			args.data.hash = temp.hash;
+
+			if (hashmap_rp_debug)
+				swaps++;
+
+			dist = existing_dist;
+		}
+
+		/*
+		 * Case 4: increment slot number and probe counter, keep going
+		 * to find free slot
+		 */
+		args.pos = increment_pos(hashmap, args.pos);
+		dist += 1;
+	}
+	fprintf(stderr, "insertion requires too many swaps\n");
+	pmemobj_cancel(pop, args.actv, args.actv_cnt);
+
+	return -1;
+}
+
+/*
+ * index_lookup -- checks if given key exists in hashmap.
+ * Returns index number if key was found, 0 otherwise.
+ */
+static uint64_t
+index_lookup(const struct hashmap_rp *hashmap, uint64_t key)
+{
+	const uint64_t hash_lookup = hash(hashmap, key);
+	uint64_t pos = hash_lookup;
+	uint64_t dist = 0;
+
+	const struct entry *entry_p = NULL;
+	while (1) {
+		entry_p = (struct entry *)pmemobj_direct(hashmap->entries.oid);
+		entry_p += pos;
+
+		if (entry_p->hash == 0 || dist > probe_distance(hashmap,
+							entry_p->hash, pos))
+			return 0;
+		else if (entry_p->hash == hash_lookup && entry_p->key == key)
+			return pos;
+
+		pos = increment_pos(hashmap, pos);
+		dist++;
+	}
+}
+
+/*
+ * entries_cache -- cache entries from first argument in entries from second
+ * argument
+ */
+static int
+entries_cache(PMEMobjpool *pop, const struct hashmap_rp *from,
```

Comment:
`struct hashmap_rp *dest, const struct hashmap_rp *src`
In this exact order.

And I don't like the name.